qupath-core-processing/src/main/java/qupath/opencv/tools/OpenCVTools.java

```diff
@@ -2651,6 +2651,70 @@ public static List<IndexedPixel> getMaskedPixels(Mat mat, Mat mask) {
 		}
 		return pixels;
 	}
+
+	/**
+	 * <p>Extract pixels from an image using a mask, limited to the specified channel index.</p>
+	 * <p>The mask can have one or multiple channels:</p>
+	 * <ul>
+	 *     <li>If the mask has only one channel, this channel is used.</li>
+	 *     <li>If the mask is multichannel, only the specified channel is used.</li>
+	 * </ul>
+	 * <p>The input and the mask must have the same dimensions.</p>
+	 *
+	 * @param input  the image from where the pixels should be extracted
+	 * @param mask  the mask to apply to the input
+	 * @param channel  the channel index (zero-based) to consider
+	 * @return a float array containing the values of the input filtered by the mask
+	 * @throws IllegalArgumentException when the input and the mask don't have the same dimensions
+	 */
+	public static float[] extractMaskedPixels(Mat input, Mat mask, int channel) {
+		Mat inputChannel = new Mat();
+		opencv_core.extractChannel(input, inputChannel, channel);
+
+		Mat maskChannel = new Mat();
+		if (mask.channels() == 1) {
+			maskChannel = mask;
+		} else {
+			opencv_core.extractChannel(mask, maskChannel, channel);
+		}
+
+		if (inputChannel.rows() != maskChannel.rows()) {
+			throw new IllegalArgumentException("The input and the mask don't have the same number of rows");
+		}
+		if (inputChannel.cols() != maskChannel.cols()) {
+			throw new IllegalArgumentException("The input and the mask don't have the same number of columns");
+		}
+
+		int height = inputChannel.rows();
+		int width = inputChannel.cols();
+		List<Float> outputList = new ArrayList<>();
+
+		try (var idx = inputChannel.createIndexer()) {
+			try (var idxMask = maskChannel.createIndexer()) {
+				for (int y = 0; y < height; y++) {
+					for (int x = 0; x < width; x++) {
+						if (idxMask.getDouble(y, x) != 0) {
```

Comment:
Assuming this calls [`Indexer.getDouble(long...)`](https://javadoc.io/doc/org.bytedeco/javacpp/latest/index.html) (and not some overridden method) then it could be optimized further by preallocating a `long[]` and updating the `y` and `x` values within the loop.

The reason is that I expect [each call to a method with a `varargs` to be slower](https://stackoverflow.com/questions/2426455/javas-varargs-performance), since an array will need to be created. Applying this twice for each pixel might end up being a long time, whereas if you can pass a single `long[]` created in advance then it should be faster.